Go 1.13  and 1.15 support with modules, updated examples and documentation (despite the branch name mentioning 1.14)

sciabarracom opened the pull request:
This PR includes support for Go 1.13 and 1.14 including many other improvements.
And supports Go Modules.

It should be functional, you can test it  but do not merge until I updated examples and documentation and removed older runtimes.

rabbah commented on CHANGES.md:
```diff
@@ -16,7 +16,10 @@
 #
 -->
-# 1.16.0 (next release)
+# 1.16.0
```
```suggestion
# 1.16.0 (next release)
```

style95 commented on docs/BUILD.md:
```diff
@@ -43,10 +43,11 @@ To build the docker images, after compiling go proxy:
-* `action-golang-v1.11`: an image supporting  Go sources
-* `actionloop`: the base image, supporting generic executables ans shell script
+* `action-golang-v1.13`: an image supporting Go 1.13 sources (does expect an ack)
+* `action-golang-v1.14`: an image supporting Go 1.14 sources (does expect an ack)
```
Should we mention v1.15 as well?

style95 commented on docs/ACTION.md:
```diff
@@ -22,7 +22,7 @@
 ## How to write Go Actions
 
-The `action-golang-v1.11` runtime can execute actions written in the Go programming language in OpenWhisk, either precompiled binary or compiling sources on the fly.
+The `action-golang-v1.14` runtime can execute actions written in the Go programming language in OpenWhisk, either precompiled binary or compiling sources on the fly.
```
Would it be better to change this to v1.15?

style95 commented on docs/DEPLOY.md:
```diff
@@ -49,120 +49,41 @@ The runtime `action-golang-v1.11` accepts:
 ## Precompiling Go Sources Offline
 
-Compiling sources on the image can take some time when the images is initialized. You can speed up precompiling the sources using the image `action-golang-v1.11` as an offline compiler. You need `docker` for doing that.
+Compiling sources on the image can take some time when the images is initialized. You can speed up precompiling the sources using the image `action-golang-v1.14` as an offline compiler. You need `docker` for doing that.
```
Same with above.

style95 commented on examples/EXAMPLES.md:
```diff
@@ -22,29 +22,24 @@
 
-- dep version 0.5.0 (check the version `dep version`)
-- Go version 1.11.1
+- Go version 1.14.6
```
a nit

style95 commented on examples/module-main/go.mod:
```diff
@@ -0,0 +1,5 @@
+module action
+
+go 1.14
```
a nit

style95 commented on examples/package-main/go.mod:
```diff
@@ -0,0 +1,7 @@
+module action
+
+go 1.14
```
a nit

style95 commented on examples/package-main/hello/go.mod:
```diff
@@ -0,0 +1,3 @@
+module hello
+
+go 1.14
```
a nit

style95 commented on golang1.14/Dockerfile:
```diff
@@ -14,27 +14,28 @@
-	docker kill under-test
-
-.PHONY: all golang bash
+FROM golang:1.13.8
```
Should this be some 1.14.x version?

sciabarracom replied:
ouch! How I did miss this?

style95 reviewed (commented):
If we start supporting go-1.15, how about updating documents accordingly too?

mrutkows commented:
@sciabarracom  I have been warned away from using go 1.14.x in any production setting due to EINTR issues...  the advice is to skip from 1.13 to 1.15 (where the comm. addressed regressions in sys libs.) which thankfully was just released a couple of weeks ago.

mrutkows commented:
@sciabarra The issue is ack'ed in the 1.14 release notes:
https://golang.org/doc/go1.14

Specifically:
"Goroutines are now asynchronously preemptible. As a result, loops without function calls no longer potentially deadlock the scheduler or significantly delay garbage collection. This is supported on all platforms except windows/arm, darwin/arm, js/wasm, and plan9/*.

A consequence of the implementation of preemption is that on Unix systems, including Linux and macOS systems, programs built with Go 1.14 will receive more signals than programs built with earlier releases. This means that programs that use packages like syscall or golang.org/x/sys/unix will see more slow system calls fail with EINTR errors. Those programs will have to handle those errors in some way, most likely looping to try the system call again. For more information about this see man 7 signal for Linux systems or similar documentation for other systems. "

Under production loads, many sys. level calls will fail because they are not coded to be interrupt tolerant (i.e., code their own retries on failure) resulting higher order errors.

sciabarracom commented:
Yes but what we do? Do we get rid of all the runtimes older than 1.15? Or just skip 1.14? 

-- 
  Michele Sciabarra
  michele@sciabarra.com



----- Original message -----
From: Matt Rutkowski <notifications@github.com>
To: apache/openwhisk-runtime-go <openwhisk-runtime-go@noreply.github.com>
Cc: Michele Sciabarra <sciabarracom@sciabarra.com>, Mention <mention@noreply.github.com>
Subject: Re: [apache/openwhisk-runtime-go] Go 1.13 1.14 and 1.15 support with modules, updated examples and documentation. (#128)
Date: Monday, August 24, 2020 11:57 PM




@sciabarra <https://github.com/sciabarra> The issue is ack'ed in the 1.14 release notes:
https://golang.org/doc/go1.14


Specifically:
"Goroutines are now asynchronously preemptible. As a result, loops without function calls no longer potentially deadlock the scheduler or significantly delay garbage collection. This is supported on all platforms except windows/arm, darwin/arm, js/wasm, and plan9/*.

A consequence of the implementation of preemption is that on Unix systems, including Linux and macOS systems, programs built with Go 1.14 will receive more signals than programs built with earlier releases. This means that programs that use packages like syscall or golang.org/x/sys/unix will see more slow system calls fail with EINTR errors. Those programs will have to handle those errors in some way, most likely looping to try the system call again. For more information about this see man 7 signal for Linux systems or similar documentation for other systems. "

Under production loads, many sys. level calls will fail because they are not coded to be interrupt tolerant (i.e., code their own retries on failure) resulting higher order errors.


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/apache/openwhisk-runtime-go/pull/128#issuecomment-679387166>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AHJ4D3YETUV4QDCLRX7BI43SCLO35ANCNFSM4QBO6K2A>.

mrutkows commented:
Hi Michele,

You must have missed the comment I made previously above the one you 
copied below.  It reads:

"@sciabarracom I have been warned away from using go 1.14.x in any 
production setting due to EINTR issues... the advice is to skip from 1.13 
to 1.15 (where the comm. addressed regressions in sys libs.) which 
thankfully was just released a couple of weeks ago."




From:   Michele Sciabarra <notifications@github.com>
To:     apache/openwhisk-runtime-go 
<openwhisk-runtime-go@noreply.github.com>
Cc:     Matt Rutkowski <mrutkows@us.ibm.com>, Comment 
<comment@noreply.github.com>
Date:   08/25/2020 02:42 AM
Subject:        [EXTERNAL] Re: [apache/openwhisk-runtime-go] Go 1.13 1.14 
and 1.15 support with modules, updated examples and documentation. (#128)



Yes but what we do? Do we get rid of all the runtimes older than 1.15? Or 
just skip 1.14? 

-- 
Michele Sciabarra
michele@sciabarra.com



----- Original message -----
From: Matt Rutkowski <notifications@github.com>
To: apache/openwhisk-runtime-go <openwhisk-runtime-go@noreply.github.com>
Cc: Michele Sciabarra <sciabarracom@sciabarra.com>, Mention 
<mention@noreply.github.com>
Subject: Re: [apache/openwhisk-runtime-go] Go 1.13 1.14 and 1.15 support 
with modules, updated examples and documentation. (#128)
Date: Monday, August 24, 2020 11:57 PM




@sciabarra <https://github.com/sciabarra> The issue is ack'ed in the 1.14 
release notes:
https://golang.org/doc/go1.14


Specifically:
"Goroutines are now asynchronously preemptible. As a result, loops without 
function calls no longer potentially deadlock the scheduler or 
significantly delay garbage collection. This is supported on all platforms 
except windows/arm, darwin/arm, js/wasm, and plan9/*.

A consequence of the implementation of preemption is that on Unix systems, 
including Linux and macOS systems, programs built with Go 1.14 will 
receive more signals than programs built with earlier releases. This means 
that programs that use packages like syscall or golang.org/x/sys/unix will 
see more slow system calls fail with EINTR errors. Those programs will 
have to handle those errors in some way, most likely looping to try the 
system call again. For more information about this see man 7 signal for 
Linux systems or similar documentation for other systems. "

Under production loads, many sys. level calls will fail because they are 
not coded to be interrupt tolerant (i.e., code their own retries on 
failure) resulting higher order errors.


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <
https://github.com/apache/openwhisk-runtime-go/pull/128#issuecomment-679387166>
, or unsubscribe <
https://github.com/notifications/unsubscribe-auth/AHJ4D3YETUV4QDCLRX7BI43SCLO35ANCNFSM4QBO6K2A>
.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or unsubscribe.

sciabarra commented:
I was asking if we have to keep 1.13

-- 
  Michele Sciabarra
  github@sciabarra.com



----- Original message -----
From: Matt Rutkowski <notifications@github.com>
To: apache/openwhisk-runtime-go <openwhisk-runtime-go@noreply.github.com>
Cc: Michele Sciabarra <github@sciabarra.com>, Mention <mention@noreply.github.com>
Subject: Re: [apache/openwhisk-runtime-go] Go 1.13 1.14 and 1.15 support with modules, updated examples and documentation. (#128)
Date: Tuesday, August 25, 2020 3:28 PM



Hi Michele,

You must have missed the comment I made previously above the one you 
copied below. It reads: 

"@sciabarracom I have been warned away from using go 1.14.x in any 
production setting due to EINTR issues... the advice is to skip from 1.13 
to 1.15 (where the comm. addressed regressions in sys libs.) which 
thankfully was just released a couple of weeks ago." 




From: Michele Sciabarra <notifications@github.com> 
To: apache/openwhisk-runtime-go 
<openwhisk-runtime-go@noreply.github.com> 
Cc: Matt Rutkowski <mrutkows@us.ibm.com>, Comment 
<comment@noreply.github.com> 
Date: 08/25/2020 02:42 AM 
Subject: [EXTERNAL] Re: [apache/openwhisk-runtime-go] Go 1.13 1.14 
and 1.15 support with modules, updated examples and documentation. (#128) 



Yes but what we do? Do we get rid of all the runtimes older than 1.15? Or 
just skip 1.14? 

-- 
Michele Sciabarra 
michele@sciabarra.com 



----- Original message ----- 
From: Matt Rutkowski <notifications@github.com> 
To: apache/openwhisk-runtime-go <openwhisk-runtime-go@noreply.github.com> 
Cc: Michele Sciabarra <sciabarracom@sciabarra.com>, Mention 
<mention@noreply.github.com> 
Subject: Re: [apache/openwhisk-runtime-go] Go 1.13 1.14 and 1.15 support 
with modules, updated examples and documentation. (#128) 
Date: Monday, August 24, 2020 11:57 PM 




@sciabarra <https://github.com/sciabarra> The issue is ack'ed in the 1.14 
release notes: 
https://golang.org/doc/go1.14 


Specifically: 
"Goroutines are now asynchronously preemptible. As a result, loops without 
function calls no longer potentially deadlock the scheduler or 
significantly delay garbage collection. This is supported on all platforms 
except windows/arm, darwin/arm, js/wasm, and plan9/*. 

A consequence of the implementation of preemption is that on Unix systems, 
including Linux and macOS systems, programs built with Go 1.14 will 
receive more signals than programs built with earlier releases. This means 
that programs that use packages like syscall or golang.org/x/sys/unix will 
see more slow system calls fail with EINTR errors. Those programs will 
have to handle those errors in some way, most likely looping to try the 
system call again. For more information about this see man 7 signal for 
Linux systems or similar documentation for other systems. " 

Under production loads, many sys. level calls will fail because they are 
not coded to be interrupt tolerant (i.e., code their own retries on 
failure) resulting higher order errors. 


— 
You are receiving this because you were mentioned. 
Reply to this email directly, view it on GitHub < 
https://github.com/apache/openwhisk-runtime-go/pull/128#issuecomment-679387166> 
, or unsubscribe < 
https://github.com/notifications/unsubscribe-auth/AHJ4D3YETUV4QDCLRX7BI43SCLO35ANCNFSM4QBO6K2A> 
. 

— 
You are receiving this because you commented. 
Reply to this email directly, view it on GitHub, or unsubscribe. 





—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/apache/openwhisk-runtime-go/pull/128#issuecomment-680025019>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAEEPCJR5HK6MJ2AYFOQBUTSCO4ANANCNFSM4QBO6K2A>.

mrutkows commented:
Had not formulated an opinion on 1.13 strongly until now; however, with 
1.15 being so new and 1.14 being skipped, the 1.13 revisions do include 
some interesting and sig. fixes and additions including many security and 
crypto fixes.  Could see some demand the 1.13 runtimes in addition as my 
view it is the last stable minor revision that still includes support for 
older CPUs and 32-bit as well. 

If it were me, I would mint a 1.13.15 runtime and then work on 1.15.x and 
if someone wants a 1.14 for some reason they can add it.




From:   Michele Sciabarra <notifications@github.com>
To:     apache/openwhisk-runtime-go 
<openwhisk-runtime-go@noreply.github.com>
Cc:     Matt Rutkowski <mrutkows@us.ibm.com>, Comment 
<comment@noreply.github.com>
Date:   08/25/2020 09:03 AM
Subject:        [EXTERNAL] Re: [apache/openwhisk-runtime-go] Go 1.13 1.14 
and 1.15 support with modules, updated examples and documentation. (#128)



I was asking if we have to keep 1.13

-- 
Michele Sciabarra
github@sciabarra.com



----- Original message -----
From: Matt Rutkowski <notifications@github.com>
To: apache/openwhisk-runtime-go <openwhisk-runtime-go@noreply.github.com>
Cc: Michele Sciabarra <github@sciabarra.com>, Mention 
<mention@noreply.github.com>
Subject: Re: [apache/openwhisk-runtime-go] Go 1.13 1.14 and 1.15 support 
with modules, updated examples and documentation. (#128)
Date: Tuesday, August 25, 2020 3:28 PM



Hi Michele,

You must have missed the comment I made previously above the one you 
copied below. It reads: 

"@sciabarracom I have been warned away from using go 1.14.x in any 
production setting due to EINTR issues... the advice is to skip from 1.13 
to 1.15 (where the comm. addressed regressions in sys libs.) which 
thankfully was just released a couple of weeks ago." 




From: Michele Sciabarra <notifications@github.com> 
To: apache/openwhisk-runtime-go 
<openwhisk-runtime-go@noreply.github.com> 
Cc: Matt Rutkowski <mrutkows@us.ibm.com>, Comment 
<comment@noreply.github.com> 
Date: 08/25/2020 02:42 AM 
Subject: [EXTERNAL] Re: [apache/openwhisk-runtime-go] Go 1.13 1.14 
and 1.15 support with modules, updated examples and documentation. (#128) 



Yes but what we do? Do we get rid of all the runtimes older than 1.15? Or 
just skip 1.14? 

-- 
Michele Sciabarra 
michele@sciabarra.com 



----- Original message ----- 
From: Matt Rutkowski <notifications@github.com> 
To: apache/openwhisk-runtime-go <openwhisk-runtime-go@noreply.github.com> 
Cc: Michele Sciabarra <sciabarracom@sciabarra.com>, Mention 
<mention@noreply.github.com> 
Subject: Re: [apache/openwhisk-runtime-go] Go 1.13 1.14 and 1.15 support 
with modules, updated examples and documentation. (#128) 
Date: Monday, August 24, 2020 11:57 PM 




@sciabarra <https://github.com/sciabarra> The issue is ack'ed in the 1.14 
release notes: 
https://golang.org/doc/go1.14 


Specifically: 
"Goroutines are now asynchronously preemptible. As a result, loops without 

function calls no longer potentially deadlock the scheduler or 
significantly delay garbage collection. This is supported on all platforms 

except windows/arm, darwin/arm, js/wasm, and plan9/*. 

A consequence of the implementation of preemption is that on Unix systems, 

including Linux and macOS systems, programs built with Go 1.14 will 
receive more signals than programs built with earlier releases. This means 

that programs that use packages like syscall or golang.org/x/sys/unix will 

see more slow system calls fail with EINTR errors. Those programs will 
have to handle those errors in some way, most likely looping to try the 
system call again. For more information about this see man 7 signal for 
Linux systems or similar documentation for other systems. " 

Under production loads, many sys. level calls will fail because they are 
not coded to be interrupt tolerant (i.e., code their own retries on 
failure) resulting higher order errors. 


— 
You are receiving this because you were mentioned. 
Reply to this email directly, view it on GitHub < 
https://github.com/apache/openwhisk-runtime-go/pull/128#issuecomment-679387166> 

, or unsubscribe < 
https://github.com/notifications/unsubscribe-auth/AHJ4D3YETUV4QDCLRX7BI43SCLO35ANCNFSM4QBO6K2A> 

. 

— 
You are receiving this because you commented. 
Reply to this email directly, view it on GitHub, or unsubscribe. 





—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <
https://github.com/apache/openwhisk-runtime-go/pull/128#issuecomment-680025019>
, or unsubscribe <
https://github.com/notifications/unsubscribe-auth/AAEEPCJR5HK6MJ2AYFOQBUTSCO4ANANCNFSM4QBO6K2A>
.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or unsubscribe.

sciabarracom commented:
I actually the runtimes are ready  with a PR pending. There is still one test failing because tests does not include a proper go.mod but it is fault of the tests not of the runtime. Will fix the test, leaving 1.13 and 1.15. I was asking in order to complete the PR.


-- 
  Michele Sciabarra
  michele@sciabarra.com



----- Original message -----
From: Matt Rutkowski <notifications@github.com>
To: apache/openwhisk-runtime-go <openwhisk-runtime-go@noreply.github.com>
Cc: Michele Sciabarra <sciabarracom@sciabarra.com>, Mention <mention@noreply.github.com>
Subject: Re: [apache/openwhisk-runtime-go] Go 1.13 1.14 and 1.15 support with modules, updated examples and documentation. (#128)
Date: Tuesday, August 25, 2020 5:08 PM



Had not formulated an opinion on 1.13 strongly until now; however, with
1.15 being so new and 1.14 being skipped, the 1.13 revisions do include 
some interesting and sig. fixes and additions including many security and 
crypto fixes. Could see some demand the 1.13 runtimes in addition as my 
view it is the last stable minor revision that still includes support for 
older CPUs and 32-bit as well. 

If it were me, I would mint a 1.13.15 runtime and then work on 1.15.x and 
if someone wants a 1.14 for some reason they can add it. 




From: Michele Sciabarra <notifications@github.com> 
To: apache/openwhisk-runtime-go 
<openwhisk-runtime-go@noreply.github.com> 
Cc: Matt Rutkowski <mrutkows@us.ibm.com>, Comment 
<comment@noreply.github.com> 
Date: 08/25/2020 09:03 AM 
Subject: [EXTERNAL] Re: [apache/openwhisk-runtime-go] Go 1.13 1.14 
and 1.15 support with modules, updated examples and documentation. (#128) 



I was asking if we have to keep 1.13 

-- 
Michele Sciabarra 
github@sciabarra.com 



----- Original message ----- 
From: Matt Rutkowski <notifications@github.com> 
To: apache/openwhisk-runtime-go <openwhisk-runtime-go@noreply.github.com> 
Cc: Michele Sciabarra <github@sciabarra.com>, Mention 
<mention@noreply.github.com> 
Subject: Re: [apache/openwhisk-runtime-go] Go 1.13 1.14 and 1.15 support 
with modules, updated examples and documentation. (#128) 
Date: Tuesday, August 25, 2020 3:28 PM 



Hi Michele, 

You must have missed the comment I made previously above the one you 
copied below. It reads: 

"@sciabarracom I have been warned away from using go 1.14.x in any 
production setting due to EINTR issues... the advice is to skip from 1.13 
to 1.15 (where the comm. addressed regressions in sys libs.) which 
thankfully was just released a couple of weeks ago." 




From: Michele Sciabarra <notifications@github.com> 
To: apache/openwhisk-runtime-go 
<openwhisk-runtime-go@noreply.github.com> 
Cc: Matt Rutkowski <mrutkows@us.ibm.com>, Comment 
<comment@noreply.github.com> 
Date: 08/25/2020 02:42 AM 
Subject: [EXTERNAL] Re: [apache/openwhisk-runtime-go] Go 1.13 1.14 
and 1.15 support with modules, updated examples and documentation. (#128) 



Yes but what we do? Do we get rid of all the runtimes older than 1.15? Or 
just skip 1.14? 

-- 
Michele Sciabarra 
michele@sciabarra.com 



----- Original message ----- 
From: Matt Rutkowski <notifications@github.com> 
To: apache/openwhisk-runtime-go <openwhisk-runtime-go@noreply.github.com> 
Cc: Michele Sciabarra <sciabarracom@sciabarra.com>, Mention 
<mention@noreply.github.com> 
Subject: Re: [apache/openwhisk-runtime-go] Go 1.13 1.14 and 1.15 support 
with modules, updated examples and documentation. (#128) 
Date: Monday, August 24, 2020 11:57 PM 




@sciabarra <https://github.com/sciabarra> The issue is ack'ed in the 1.14 
release notes: 
https://golang.org/doc/go1.14 


Specifically: 
"Goroutines are now asynchronously preemptible. As a result, loops without 

function calls no longer potentially deadlock the scheduler or 
significantly delay garbage collection. This is supported on all platforms 

except windows/arm, darwin/arm, js/wasm, and plan9/*. 

A consequence of the implementation of preemption is that on Unix systems, 

including Linux and macOS systems, programs built with Go 1.14 will 
receive more signals than programs built with earlier releases. This means 

that programs that use packages like syscall or golang.org/x/sys/unix will 

see more slow system calls fail with EINTR errors. Those programs will 
have to handle those errors in some way, most likely looping to try the 
system call again. For more information about this see man 7 signal for 
Linux systems or similar documentation for other systems. " 

Under production loads, many sys. level calls will fail because they are 
not coded to be interrupt tolerant (i.e., code their own retries on 
failure) resulting higher order errors. 


— 
You are receiving this because you were mentioned. 
Reply to this email directly, view it on GitHub < 
https://github.com/apache/openwhisk-runtime-go/pull/128#issuecomment-679387166> 

, or unsubscribe < 
https://github.com/notifications/unsubscribe-auth/AHJ4D3YETUV4QDCLRX7BI43SCLO35ANCNFSM4QBO6K2A> 

. 

— 
You are receiving this because you commented. 
Reply to this email directly, view it on GitHub, or unsubscribe. 





— 
You are receiving this because you were mentioned. 
Reply to this email directly, view it on GitHub < 
https://github.com/apache/openwhisk-runtime-go/pull/128#issuecomment-680025019> 
, or unsubscribe < 
https://github.com/notifications/unsubscribe-auth/AAEEPCJR5HK6MJ2AYFOQBUTSCO4ANANCNFSM4QBO6K2A> 
. 

— 
You are receiving this because you commented. 
Reply to this email directly, view it on GitHub, or unsubscribe. 





—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/apache/openwhisk-runtime-go/pull/128#issuecomment-680083829>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AHJ4D33PNXWYH22FDY3XYNDSCPHVDANCNFSM4QBO6K2A>.
